### PR TITLE
Feature/training script

### PR DESCRIPTION
**Pull request by jarzeckil**
- wrapper for dataset (for albumentations)
- factory for loaders and transforms
- model wrapper - training and eval
- basic training script
- hydra config for object instantation

### CODE REVIEW

File: ai_detector_model/modeling/train.py [hunk 1]
```diff
@@ -31,12 +35,54 @@
...
+
+    model = model.to(device=DEVICE)
+
+    loss_fn = hydra.utils.instantiate(cfg.train.loss_function)
+    optim = optim_partial(
+        [
+            {"params": backbone_params, "lr": base_lr * 0.1},
+            {"params": head_params},
+        ]
+    )
```
[the remark this answers is not in the document]
Comment (jarzeckil) [reply ai_detector_model/modeling/train.py:87]: we won't be using baseline model ig

File: ai_detector_model/modeling/engine.py [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+
+class Trainer:
+    def __init__(
+        self,
+        model: nn.Module,
+        loss_fn: nn.Module,
+        optim: Optimizer,
+        train_loader: DataLoader | None,
+        test_loader: DataLoader | None,
+        device: str,
```
[the remark this answers is not in the document]
Comment (jarzeckil) [reply ai_detector_model/modeling/engine.py:24]: done

File: ai_detector_model/data/factory.py [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+
+
+def create_transforms(cfg: DictConfig, stage="train") -> A.Compose:
+    image_size = cfg.model.input_size
+    tr_cfg = cfg.data.dataset.transforms
+    if stage == "train":
+        return A.Compose(
+            [
+                A.Resize(image_size, image_size),
+                A.HorizontalFlip(tr_cfg.horizontal_flip_p),
```
[the remark this answers is not in the document]
Comment (jarzeckil) [reply ai_detector_model/data/factory.py]: done

File: ai_detector_model/data/factory.py [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+
+    train_dataset = ClassificationDataset(train_path, train_transform)
+    test_dataset = ClassificationDataset(test_path, test_transform)
+
+    train_loader = DataLoader(
+        dataset=train_dataset,
+        batch_size=cfg.data.dataset.batch_size,
+        num_workers=cfg.data.dataset.num_workers,
+        pin_memory=True,
+        shuffle=False,
```
[the remark this answers is not in the document]
Comment (jarzeckil) [reply ai_detector_model/data/factory.py]: done

File: ai_detector_model/data/factory.py [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+    )
+    test_loader = DataLoader(
+        dataset=test_dataset,
+        batch_size=cfg.data.dataset.batch_size,
+        num_workers=cfg.data.dataset.num_workers,
+        pin_memory=True,
```
[the remark this answers is not in the document]
Comment (jarzeckil) [reply ai_detector_model/data/factory.py]: done

File: ai_detector_model/data/factory.py [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+def create_loaders(cfg: DictConfig) -> tuple[DataLoader, DataLoader]:
+    train_transform = create_transforms(cfg, "train")
+    test_transform = create_transforms(cfg, "test")
+
+    train_path = os.path.join(PROCESSED_DATA_DIR, "train")
+    test_path = os.path.join(PROCESSED_DATA_DIR, "test")
+
+    train_dataset = ClassificationDataset(train_path, train_transform)
+    test_dataset = ClassificationDataset(test_path, test_transform)
+
```
[the remark this answers is not in the document]
Comment (jarzeckil) [reply ai_detector_model/data/factory.py]: done

**Review by Janek0337 — APPROVED** [review]
valid

File: ai_detector_model/modeling/engine.py [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+                losses.append(loss.item())
+
+                self.metrics.update(torch.sigmoid(pred), labels)
+
+        scores = {k: v.item() for k, v in self.metrics.compute().items()}
+        scores["validation_loss"] = np.mean(losses)
+
+        return scores
+
+    def train(self, max_epochs: int):
```
Comment (JGolaszewski) [inline ai_detector_model/modeling/engine.py:108]: This is not okay since you can specify any metric for eval. If f1 is not used this will not work.

File: ai_detector_model/data/factory.py [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+        )
+    raise ValueError("Wrong stage has been specified")
+
+
+def create_train_loader(cfg: DictConfig) -> DataLoader[Any]:
+    train_transform = create_transforms(cfg, "train")
+    train_path = os.path.join(cfg.data.dataset.path, "train")
+    train_dataset = ClassificationDataset(train_path, train_transform)
+
+    train_loader = DataLoader(
```
Comment (JGolaszewski) [inline ai_detector_model/data/factory.py:41]: Add prefetch factor

File: ai_detector_model/modeling/engine.py [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+
+        losses = []
+        images: Tensor
+        labels: Tensor
+        self.optim.zero_grad()
+        for i, (images, labels) in enumerate(self.train_loader):
+            images = images.to(self.device)
+            labels = labels.to(self.device)
+
+            pred = self.model(images).squeeze(-1)
```
Comment (JGolaszewski) [inline ai_detector_model/modeling/engine.py]: I would also add option to enable torch amp for faster training

**Review by JGolaszewski — CHANGES_REQUESTED** [review]
Overall really good small changes needed in the comments.

Things to consider for next task:
- if we will use some heavy augmentation things we may need to create option to do them in training loop in gpu but this depends on you and Pawel
- you may want to have option to use lr schedulers for refined training

File: ai_detector_model/modeling/engine.py (hunk 9 above)
Comment (JGolaszewski) [reply ai_detector_model/modeling/engine.py]: Some of the things are as comments idk why

File: ai_detector_model/modeling/engine.py [hunk 10]
```diff
@@ -0,0 +1,143 @@
...
+        losses = []
+        images: Tensor
+        labels: Tensor
+        for i, (images, labels) in enumerate(self.train_loader):
+            self.optim.zero_grad()
+            images = images.to(self.device, non_blocking=True)
+            labels = labels.to(self.device, non_blocking=True)
+
+            with torch.autocast(
+                device_type=self.device, dtype=torch.float16, enabled=self.use_amp
```
Comment (JGolaszewski) [inline ai_detector_model/modeling/engine.py]: For mixed precision training it is recommended to use gradient scaling if type is float16 (in pytorch `GradScaler`).

You can either use float16 with `GradScaler` or bfloat16 without.

https://docs.pytorch.org/tutorials/recipes/recipes/amp_recipe.html#automatic-mixed-precision

**Review by JGolaszewski — CHANGES_REQUESTED** [review]
Very nice work, one small thing to change
